cli/admin/src/test/java/io/pravega/cli/admin/dataRecovery/DataRecoveryTest.java

```diff
@@ -0,0 +1,490 @@
+            }
+            this.serviceBuilder.initialize();
+            this.streamSegmentStore = this.serviceBuilder.createStreamSegmentService();
+            this.tableStore = this.serviceBuilder.createTableStoreService();
+            this.server = new PravegaConnectionListener(false, servicePort, this.streamSegmentStore, this.tableStore,
+                    this.serviceBuilder.getLowPriorityExecutor());
+            this.server.startListening();
+        }
+
+        @Override
+        public void close() {
+            this.server.close();
+            this.serviceBuilder.close();
+        }
+    }
+
+    /**
+     * Creates a controller instance and runs it.
+     */
+    private static class ControllerRunner implements AutoCloseable {
+        private final int controllerPort = io.pravega.test.common.TestUtils.getAvailableListenPort();
+        private final String serviceHost = "localhost";
+        private final ControllerWrapper controllerWrapper;
+        private final Controller controller;
+        private final URI controllerURI = URI.create("tcp://" + serviceHost + ":" + controllerPort);
+
+        ControllerRunner(int bkPort, int servicePort, int containerCount) throws InterruptedException {
+            this.controllerWrapper = new ControllerWrapper("localhost:" + bkPort, false,
+                    controllerPort, serviceHost, servicePort, containerCount);
+            this.controllerWrapper.awaitRunning();
+            this.controller = controllerWrapper.getController();
+        }
+
+        @Override
+        public void close() throws Exception {
+            this.controller.close();
+            this.controllerWrapper.close();
+        }
+    }
+
+    /**
+     * Creates a client to read and write events.
+     */
+    private static class ClientRunner implements AutoCloseable {
+        private final ConnectionFactory connectionFactory;
+        private final ClientFactoryImpl clientFactory;
+        private final ReaderGroupManager readerGroupManager;
+
+        ClientRunner(ControllerRunner controllerRunner) {
+            this.connectionFactory = new SocketConnectionFactoryImpl(ClientConfig.builder()
+                    .controllerURI(controllerRunner.controllerURI).build());
+            this.clientFactory = new ClientFactoryImpl(SCOPE, controllerRunner.controller, connectionFactory);
+            this.readerGroupManager = new ReaderGroupManagerImpl(SCOPE, controllerRunner.controller, clientFactory);
+        }
+
+        @Override
+        public void close() {
+            this.readerGroupManager.close();
+            this.clientFactory.close();
+            this.connectionFactory.close();
+        }
+    }
+
+    /**
+     * Creates a Pravega instance.
+     */
+    private static class PravegaRunner implements AutoCloseable {
+        private final int containerCount;
+        private BookKeeperRunner bookKeeperRunner;
```

Comment:
This one can be made final.

Reply:
No. It can't be. I am reassigning it to clear zk/bk and then start a new one.

Reply:
This is a real bad practice. If you let an external class tinker with your private fields, then you are essentially breaking encapsulation, which can lead to a lot of bugs.

I suggest you make a method on this class which will properly close the existing runner and reassign it to a new one.

That way nobody else will need to know to close your instances or whatever. See below. You are closing your other runners here and also in the test above. How can you track that? If you want to debug something it will be next to impossible to figure out your program flow. 

Please respect the encapsulation rules and move all management of these runners within this class.

Reply:
Thanks for letting me know. I made some changes on other parts as well, considering this.